### ISSUE TITLE
Re-use `getCachedData` when `refresh`ing `asyncData`

### ISSUE THREAD

**Issue by TheAlexLichter** [nuxt/nuxt#24332]
### Describe the feature

Hey 👋🏻

With the introduction of `getCachedData`, `useFetch` and `useAsyncData` have an elegant way to re-use existing data easily.

There is a downside though, as @broox [mentioned in this YouTube comment](https://www.youtube.com/watch?v=aQPR0xn-MMk&lc=UgyH9sQQrLwAG9NGoEh4AaABAg):

When **refreshing the data**, either via calling `refresh` manually or implicitly using it through the `watch` property or by defining reactive parameters like `query` in the `useFetch` options, there is **no option to re-use the cached data**.

Reason behind it is that we only take cached data into account when nuxt is hydrating *or* on the initial request of the composable. https://github.com/nuxt/nuxt/blob/a087b87d3ff7f64c779199f0b4f1a866625671fa/packages/nuxt/src/app/composables/asyncData.ts#L182-L184

Spontaneous idea to resolve this issue:

1. Add a `force` parameter for `refresh`, so manual `refresh()` calls can bypass the `getCachedData` function if necessary. By default, this can be set to `true` if we want to avoid breaking changes.
3. Set `force` to `false` by default for `watch`. This is a behavior change, but when implementing an own `getCachedData` function you can take this into account easily in there. Projects not using the `getCachedData` function should not be impacted.

More than happy for other ideas or feedback ☺️

### Additional information

- [X] Would you be willing to help implement this feature?
- [ ] Could this feature be implemented as a module?

### Final checks

- [X] Read the [contribution guide](https://nuxt.com/docs/community/contribution).
- [X] Check existing [discussions](https://github.com/nuxt/nuxt/discussions) and [issues](https://github.com/nuxt/nuxt/issues).

**Comment by broox** [comment]
in my opinion, it should be default behavior that all requests, even those made by reactive changes, use the cache if your definition of `getCachedData()` resolves cached data.

**Comment by manniL** [comment]
> i.e. the definition of getCachedData() should be the "forcing" function to use the cache.

I think that sounds fair, if the function is specified and returns data, it should be used instead of calling again by default. 

But it'd be still nice to have a chance to bypass the function "on demand" (which would be equivalent to a hard reload or similar). I wonder if we need the `force` parameter for that though or if that could be implemented in userland (which I guess it could be easily. 🤔

**Comment by broox** [comment]
ah, yes. maybe it would be best if reactive changes attempt to use the cache, but `refresh()` does not?

or, like you suggested, `refresh()` could use the cache unless the `force` parameter is passed into it? similar to tapping refresh vs shift+refresh in your browser.

**Comment by darioferderber** [comment]
Hi! 

The problem is that `key` is changed.
You can try setting your own `key`.

```ts
const { data } = await useFetch('https://icanhazdadjoke.com/', {
  key: 'myOwnKey',
  query: { page },
  getCachedData: (key) => nuxt.payload.data[key] || nuxt.static.data[key],
  headers: {
    Accept: 'application/json',
  },
});
```

Hope this helps! 👋

**Comment by manniL** [comment]
> Hi!
> 
> The problem is that `key` is changed. You can try setting your own `key`.
> 
> ```ts
> const { data } = await useFetch('https://icanhazdadjoke.com/', {
>   key: 'myOwnKey',
>   query: { page },
>   getCachedData: (key) => nuxt.payload.data[key] || nuxt.static.data[key],
>   headers: {
>     Accept: 'application/json',
>   },
> });
> ```
> 
> Hope this helps! 👋

I thought so too initially but it is not enough if e.g. `query` is reactive.

**Comment by darioferderber** [comment]
> I thought so too initially but it is not enough if e.g. `query` is reactive.

Do you mean like in this [example](https://stackblitz.com/edit/nuxt-starter-rir5p2?file=pages%2Findex.vue)?

**Comment by manniL** [comment]
As soon as you use a page parameter it works, but if you "re-use" a local `useFetch` call [it is tricky](https://stackblitz.com/edit/nuxt-starter-wqhxem?file=pages%2Findex.vue)

**Comment by darioferderber** [comment]
> As soon as you use a page parameter it works, but if you "re-use" a local `useFetch` call [it is tricky](https://stackblitz.com/edit/nuxt-starter-wqhxem?file=pages%2Findex.vue)

Thanks, I get it now 🙂 

There's already an option in [asycData](https://github.com/nuxt/nuxt/blob/a087b87d3ff7f64c779199f0b4f1a866625671fa/packages/nuxt/src/app/composables/asyncData.ts#L173C54-L173C54) - `_initial`, I think it could help because it will return the cached data if you set it to `true`.

```ts
 if ((opts._initial || (nuxt.isHydrating && opts._initial !== false)) && hasCachedData()) { 
   return Promise.resolve(options.getCachedData!(key)) 
 } 
 ```

So this could be the way to "re-use" a local `useFetch`
```vue
<button @click="refresh({ _initial: true })">New Joke (refresh)</button>
```

Maybe it's a bit "hacky" but it should work 🤔

**Comment by manniL** [comment]
@darioferderber `_initial` is an *internal* app not meant for usage though (hance the `_` in the name 😋) - so yes, that might work but better DX would be nicer.

**Comment by darioferderber** [comment]
> @darioferderber `_initial` is an _internal_ app not meant for usage though (hance the `_` in the name 😋) - so yes, that might work but better DX would be nicer.

Yup, I know, but it could help @broox if he's still struggling with this issue 😄 

I agree, there should be a better DX to deal with this, but basically it should work the same as `_initial`

**Comment by broox** [comment]
@darioferderber - i don't think the private `_initial` hack helps because i'm not calling `refresh()` manually.

the http request is automatically triggered when i change the reactive `query` parameters of `useFetch()`. in this case, i'd simply like the request to respect the cached data.

**Comment by manniL** [comment]
> the http request is automatically triggered when i change the reactive `query` parameters of `useFetch()`. in this case, i'd simply like the request to respect the cached data.

(Internally, a `refresh` is called IIRC.)

**Comment by AndersCV** [comment]
Is there any update on this one? I spent quite a while thinking it was my own bug trying figuring out why my getCachedData function was only being triggered on the initial requests and not re-used when my watch sources changes. 

This would be a great addition and could save me a bunch of request with my paginated data.

**Comment by raphi-0901** [comment]
What I did as a workaround is this:
```js
const { locale } = useI18n()
const key = computed(() => `unique-key-${locale.value}`)
const nuxtApp = useNuxtApp()

const { data } = await useAsyncData(
  key.value,
  async() => {
    const cachedData = nuxtApp.payload.data[key] || nuxtApp.static.data[key]
    if (cachedData) {
      return cachedData
    }

    const response = await $fetch('https://your-fetch')

    return response.data
  },
  {
    watch: [locale],
    // getCachedData(key, nuxtApp) {
    //   console.log('nuxtApp.payload.data[key] :>>', nuxtApp.payload.data[key])
    //
    //   return nuxtApp.payload.data[key] || nuxtApp.static.data[key]
    // },
  },
)
```

It basically bypasses the getCachedData logic and does the same before fetching.

**Comment by whoisarjen** [comment]
Here something from me

```
import _ from 'lodash'
import type {  NitroFetchRequest } from 'nitropack';
import type { AsyncData } from 'nuxt/app';

export const useCustomFetch = async <T>(
  url: NitroFetchRequest,
  query: globalThis.Ref<object>,
): Promise<AsyncData<T, any>> => {
  const key = computed(() => {
    const sortedObj = _.fromPairs(_.sortBy(_.toPairs(query.value)));
    const uniqueKey = JSON.stringify(sortedObj);

    return `${url}${uniqueKey}`
  })

  return await useAsyncData(
    btoa(key.value),
    async (nuxt) => {
      if (key.value in nuxt!.payload.data) {
        return nuxt!.payload.data[key.value]
      }
  
      const result = await $fetch(url, {
        method: 'get',
        query: query.value,
      })

      nuxt!.payload.data[key.value] = result
  
      return result
    },
    {
      lazy: true,
      watch: [() => query.value],
    },
  ) as unknown as Promise<AsyncData<T, any>>
}

```

**Comment by ElYaiko** [comment]
Any update to this?, like to "share" the data state between other components consuming the same api route?
Example:
https://stackblitz.com/edit/github-xeker5?file=components%2FHelloWorld.vue

The idea is that since the data has to be fetched manually but they have the same key, when data is fetched in one component the others using the same key will share the updated data too. (In the example it almost syncs the data but throws error at the end, and without the same key the data is fetched singular)
The only current way of doing this is with a separed `useState` + a custom fetch that sets the state each time is fetched but the concept of `asyncData` to fetch data starts to gets a bit more complex because now the data has to be obtained from the `useState` instead `data` from the `asyncData` so here's when `getCachedData` could be useful like to synchronize `data` based on a `useState` so this concept of `asyncData` is simpler but currently only obtains the data at the first fetch and is not sync anymore.

**Comment by TimGuendel** [comment]
I wish this was documented. I opened [this PR](https://github.com/nuxt/nuxt/pull/29300)

**Comment by MuhammadM1998** [comment]
I've encountered the same issue when trying to figure out why a component in my app does not use the cached data when triggered by an external `watch`. I hope this gets implemented. Maybe change the default to always take `useCachedData` result into account as a breaking change in v4 🤔

**Comment by lud-hu** [comment]
Are there any plans on when to implement this?

**Comment by Gwynerva** [comment]
@danielroe This is a must have feature for SSG sites. And this is what I was trying to explain in my [recent](https://github.com/nuxt/nuxt/issues/30663#issuecomment-2602310572) issue. Right now it only calls `getCachedData` only once — on initial `useFetch`. Using `refresh` or `execute` or just changing the ref its is watching will skip looking for cached data and tries to fetch.

I suppose we need an additional option `useCache` in both `useFetch` and `useAsyncData` composables for always looking for data in payload/static first and only after that try to fetch.

To workaround this in SSG sites we have to overcomplicate logic and manually check for static payload first.

And what is even more stange, on initial page load the `nuxtApp.static.data[key]` is undefined for some reason and the actual data for this page is located in `nuxtApp.payload.data[key]`. I am not sure but this looks like a bug too.

**Comment by nikitaWebApps** [comment]
> What I did as a workaround is this:
> 
> const { locale } = useI18n()
> const key = computed(() => `unique-key-${locale.value}`)
> const nuxtApp = useNuxtApp()
> 
> const { data } = await useAsyncData(
>   key.value,
>   async() => {
>     const cachedData = nuxtApp.payload.data[key] || nuxtApp.static.data[key]
>     if (cachedData) {
>       return cachedData
>     }
> 
>     const response = await $fetch('https://your-fetch')
> 
>     return response.data
>   },
>   {
>     watch: [locale],
>     // getCachedData(key, nuxtApp) {
>     //   console.log('nuxtApp.payload.data[key] :>>', nuxtApp.payload.data[key])
>     //
>     //   return nuxtApp.payload.data[key] || nuxtApp.static.data[key]
>     // },
>   },
> )
> It basically bypasses the getCachedData logic and does the same before fetching.

It didn't quite work for me. When navigating client-side, the payload isn't updated with the new key. In my case, I specifically need to set the payload with the new key like this before returning the response:

`nuxtApp.payload.data[fetchKey.value] = response`
